### PR TITLE
Ultimate messenger API

### PR DESCRIPTION
**Pull request by JarvisCraft**
Basic APIs of `ultimate-messenger`:
- [x] `Message`
- [x] `StringFormatter`
- [x] `TextModel`
- [x] `Placeholders`
- [ ] ~Messenger~ moved to other PR
- [ ] ~MessageType~ moved to other PR
- [ ] ~MessageParser~ moved to other PR

### CODE REVIEW

File: ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+
+        /**
+         * Converts the given class name to the corresponding bytecode class name.
+         *
+         * @param className name of the class at runtime (using dot delimiter)
+         * @return name of the class in bytecode
+         *
+         * @see #toBytecodeClassName(Class)
+         */
+        protected static String toBytecodeClassName(@NonNull final String className) {
```
Comment (xdelsy) [inline ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java]: The ASM API already has such a method. See [org.objectweb.asm.Type#getType(java.lang.String)](https://asm.ow2.io/javadoc/org/objectweb/asm/Type.html#getType-java.lang.String-)

File: ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+
+        /**
+         * Gets the given class's name used in bytecode.
+         *
+         * @param clazz class whose bytecode name should be computed
+         * @return name of the class in bytecode
+         *
+         * @see #toBytecodeClassName(String)
+         */
+        protected static String toBytecodeClassName(@NonNull final Class<?> clazz) {
```
Comment (xdelsy) [inline ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java]: See [org.objectweb.asm.Type#getType(java.lang.Class<?>)](https://asm.ow2.io/javadoc/org/objectweb/asm/Type.html#getType-java.lang.Class-)

File: ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+
+                method.visitCode();
+                //<editor-fold desc="Method code generation" defaultstate="collapsed">
+                method.visitTypeInsn(NEW, STRING_BUILDER_JVM_CLASS_NAME);
+                method.visitInsn(DUP);
+                // Specify initial length of StringBuilder via its constructor
+                {
+                    val length = staticLength;
+                    switch (length) {
+                        case -1: {
```
Comment (xdelsy) [inline ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java]: Actually, the length cannot be less than 0.

File: ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java (hunk 1 above)
Comment (xdelsy) [inline ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java]: The ASM API already has such a method. See [org.objectweb.asm.Type#getType(java.lang.String)](https://asm.ow2.io/javadoc/org/objectweb/asm/Type.html#getType-java.lang.String-)

File: ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java (hunk 3 above)
Comment (JarvisCraft) [reply ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java]: Resolved

File: ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java (hunk 1 above)
Comment (JarvisCraft) [reply ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java]: Resolved in 987c4f3405191dd281905585780f08d2b5f1d13f

File: ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java (hunk 2 above)
Comment (JarvisCraft) [reply ultimate-messenger/src/main/java/ru/progrm_jarvis/ultimatemessenger/format/model/AsmGeneratingTextModelFactory.java]: Resolved in 987c4f3405191dd281905585780f08d2b5f1d13f

**Comment by JarvisCraft** [comment]
This will get merged now as it contains significant patches needed for other PRs.
As for the future, better splitting of features should happen.

**Pull request merged**